File: research/stp-v1.2/tools/concurrency.mjs
function canonical(value) {
  if (value === null || typeof value === "boolean" || typeof value === "number" || typeof value === "string") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
  return `{${Object.keys(value).sort().map((key) => `${JSON.stringify(key)}:${canonical(value[key])}`).join(",")}}`;
}

function permutations(items) {
  if (items.length <= 1) return [items];
  const result = [];
  for (let index = 0; index < items.length; index += 1) {
    const head = items[index];
    const rest = [...items.slice(0, index), ...items.slice(index + 1)];
    for (const tail of permutations(rest)) result.push([head, ...tail]);
  }
  return result;
}

function combinations(items, size) {
  if (size === 0) return [[]];
  if (items.length < size) return [];
  const [head, ...tail] = items;
  return [
    ...combinations(tail, size - 1).map((rest) => [head, ...rest]),
    ...combinations(tail, size)
  ];
}

function sameSet(left, right) {
  return canonical([...left].sort()) === canonical([...right].sort());
}

function validateDelta(profile, delta) {
  if (delta.subject_id !== profile.subject_id || delta.policy_id !== profile.policy_id || delta.policy_generation !== profile.policy_generation) {
    return "POLICY_BINDING_MISMATCH";
  }
  if (!sameSet(delta.causal_parents, profile.causal_frontier)) return "CAUSAL_FRONTIER_MISMATCH";
  return null;
}

function predicateHolds(values, predicate) {
  const present = Object.prototype.hasOwnProperty.call(values, predicate.field);
  const actual = values[predicate.field];
  if (predicate.operator === "EXISTS") return present === predicate.value;
  if (!present) return false;
  if (predicate.operator === "EQ") return canonical(actual) === canonical(predicate.value);
  if (predicate.operator === "NE") return canonical(actual) !== canonical(predicate.value);
  if (typeof actual !== "number" || typeof predicate.value !== "number") return false;
  if (predicate.operator === "LT") return actual < predicate.value;
  if (predicate.operator === "LTE") return actual <= predicate.value;
  if (predicate.operator === "GT") return actual > predicate.value;
  if (predicate.operator === "GTE") return actual >= predicate.value;
  return false;
}

function predicatesHold(values, delta) {
  return (delta.predicate_reads ?? []).every((predicate) => predicateHolds(values, predicate));
}

function runSchedule(profile, baseState, schedule) {
  const values = structuredClone(baseState.values ?? {});
  const accepted = [];
  const refused = [];
  const effectTrace = [];
  for (const delta of schedule) {
    if (!predicatesHold(values, delta)) {
      refused.push(delta.delta_id);
      continue;
    }
    const candidate = structuredClone(values);
    for (const [key, value] of Object.entries(delta.writes ?? {})) candidate[key] = value;
    for (const [key, increment] of Object.entries(delta.increments ?? {})) candidate[key] = (candidate[key] ?? 0) + increment;
    const violated = Object.entries(profile.limits)
      .filter(([, limit]) => typeof limit === "number")
      .some(([key, limit]) => typeof candidate[key] === "number" && candidate[key] > limit);
    if (violated) {
      refused.push(delta.delta_id);
      continue;
    }
    Object.assign(values, candidate);
    accepted.push(delta.delta_id);
    for (const effect of delta.effects ?? []) effectTrace.push(effect.effect_id);
  }
  return {
    projection: { values, accepted: accepted.sort(), refused: refused.sort() },
    effect_trace: effectTrace
  };
}

function scheduleClasses(profile, baseState, deltas) {
  const runs = permutations(deltas).map((schedule) => runSchedule(profile, baseState, schedule));
  return {
    schedule_count: runs.length,
    projection_count: new Set(runs.map((run) => canonical(run.projection))).size,
    effect_count: new Set(runs.map((run) => canonical(run.effect_trace))).size
  };
}

export function analyzeBatch(profile, item) {
  if (item.termination_evidence !== "FINITE_ENUMERATION") {
    return { status: "UNKNOWN", reason_code: "TERMINATION_NOT_PROVEN", schedule_count: 0, pairwise_all_commute: false };
  }
  for (const delta of item.deltas) {
    const rejection = validateDelta(profile, delta);
    if (rejection) return { status: "FAIL", reason_code: rejection, schedule_count: 0, pairwise_all_commute: false };
  }
  if (item.deltas.some((delta) => !predicatesHold(item.base_state.values ?? {}, delta))) {
    return { status: "FAIL", reason_code: "PRECONDITION_FAILED", schedule_count: 0, pairwise_all_commute: false };
  }
  const full = scheduleClasses(profile, item.base_state, item.deltas);
  const pairs = combinations(item.deltas, 2);
  const pairwiseAllCommute = pairs.every((pair) => {
    const analysis = scheduleClasses(profile, item.base_state, pair);
    return analysis.projection_count === 1 && analysis.effect_count === 1;
  });
  if (full.projection_count > 1) {
    return {
      status: "FAIL",
      reason_code: item.deltas.length > 2 && pairwiseAllCommute ? "PAIRWISE_INSUFFICIENT" : "ORDER_REQUIRED",
      schedule_count: full.schedule_count,
      pairwise_all_commute: pairwiseAllCommute
    };
  }
  if (full.effect_count > 1) {
    return { status: "FAIL", reason_code: "EFFECT_ORDER_REQUIRED", schedule_count: full.schedule_count, pairwise_all_commute: false };
  }
  return { status: "PASS", reason_code: "COMMUTES", schedule_count: full.schedule_count, pairwise_all_commute: pairwiseAllCommute };
}

export function analyzeEffect(profile, item) {
  let status = "UNKNOWN";
  let reasonCode = "NO_EVENT";
  let effectCount = 0;
  const effects = new Map();
  const keys = new Map();
  for (let index = 0; index < item.events.length; index += 1) {
    const event = item.events[index];
    const expectedSequence = index + 1;
    const expectedPrevious = index === 0 ? null : index;
    if (event.sequence !== expectedSequence || event.previous_sequence !== expectedPrevious) {
      return { status: "FAIL", reason_code: "NONCONTIGUOUS_EFFECT_SEQUENCE", effect_count: effectCount };
    }
    if (event.type !== "OBSERVE" && event.policy_generation !== profile.policy_generation) {
      return { status: "FAIL", reason_code: "POLICY_BINDING_MISMATCH", effect_count: effectCount };
    }
    const keyOwner = keys.get(event.idempotency_key);
    if (keyOwner !== undefined && keyOwner !== event.effect_id) {
      return { status: "FAIL", reason_code: "EFFECT_ID_MISMATCH", effect_count: effectCount };
    }
    let effect = effects.get(event.effect_id);
    if (!effect && event.type !== "PREPARE") {
      return {
        status: "FAIL",
        reason_code: effects.size > 0 ? "EFFECT_ID_MISMATCH" : "ILLEGAL_EFFECT_TRANSITION",
        effect_count: effectCount
      };
    }
    if (event.type === "PREPARE") {
      if (effect) return { status: "FAIL", reason_code: "ILLEGAL_EFFECT_TRANSITION", effect_count: effectCount };
      effect = { state: "PREPARED", idempotency_key: event.idempotency_key, dispatch_step: null, duplicate_pending: false, outcomes: new Set() };
      effects.set(event.effect_id, effect);
      keys.set(event.idempotency_key, event.effect_id);
      status = "UNKNOWN";
      reasonCode = "PREPARED_NOT_FINAL";
    } else if (effect.idempotency_key !== event.idempotency_key) {
      return { status: "FAIL", reason_code: "EFFECT_ID_MISMATCH", effect_count: effectCount };
    } else if (event.type === "DISPATCH") {
      if (effect.state === "PREPARED") {
        effect.state = "DISPATCHED";
        effect.dispatch_step = event.effect_step;
        effectCount += 1;
        status = "UNKNOWN";
        reasonCode = "DISPATCHED_NOT_FINAL";
      } else if (effect.state === "DISPATCHED") {
        if (event.effect_step - effect.dispatch_step <= profile.effect_policy.dedupe_retention_steps) {
          effect.duplicate_pending = true;
          status = "UNKNOWN";
          reasonCode = "DUPLICATE_SUPPRESSION_UNVERIFIED";
        } else {
          effectCount += 1;
          effect.state = "DUPLICATE_RISK";
          status = "FAIL";
          reasonCode = "DUPLICATE_RISK";
        }
      } else {
        return { status: "FAIL", reason_code: "ILLEGAL_EFFECT_TRANSITION", effect_count: effectCount };
      }
    } else if (event.type === "DEDUPE_CONFIRM") {
      if (effect.state !== "DISPATCHED" || !effect.duplicate_pending) {
        return { status: "FAIL", reason_code: "ILLEGAL_EFFECT_TRANSITION", effect_count: effectCount };
      }
      effect.duplicate_pending = false;
      status = "PASS";
      reasonCode = "DUPLICATE_DEDUPED";
    } else if (event.type === "TIMEOUT") {
      if (effect.state !== "DISPATCHED") return { status: "FAIL", reason_code: "ILLEGAL_EFFECT_TRANSITION", effect_count: effectCount };
      effect.state = "EFFECT_UNKNOWN";
      status = "UNKNOWN";
      reasonCode = "EFFECT_UNKNOWN";
    } else if (event.type === "OBSERVE") {
      if (!new Set(["DISPATCHED", "EFFECT_UNKNOWN", "OBSERVED"]).has(effect.state)) {
        return { status: "FAIL", reason_code: "ILLEGAL_EFFECT_TRANSITION", effect_count: effectCount };
      }
      if (event.policy_generation !== profile.policy_generation) {
        effect.state = "OBSERVATION_STALE";
        status = "UNKNOWN";
        reasonCode = "OBSERVATION_STALE";
      } else {
        effect.outcomes.add(event.outcome);
        if (effect.outcomes.size > 1) {
          effect.state = "OBSERVATION_CONFLICT";
          status = "UNKNOWN";
          reasonCode = "OBSERVATION_CONFLICT";
        } else {
          effect.state = "OBSERVED";
          status = "UNKNOWN";
          reasonCode = "OBSERVED_NOT_SETTLED";
        }
      }
    } else if (event.type === "SETTLE") {
      if (effect.state !== "OBSERVED") return { status: "FAIL", reason_code: "ILLEGAL_EFFECT_TRANSITION", effect_count: effectCount };
      effect.state = "SETTLED";
      status = "PASS";
      reasonCode = "SETTLED";
    } else if (event.type === "REVOKE") {
      if (effect.state === "PREPARED") {
        effect.state = "REFUSED";
        status = "FAIL";
        reasonCode = "AUTHORITY_REVOKED_BEFORE_DISPATCH";
      } else if (new Set(["DISPATCHED", "EFFECT_UNKNOWN", "OBSERVED"]).has(effect.state)) {
        effect.state = "EFFECT_UNKNOWN";
        status = "UNKNOWN";
        reasonCode = "AUTHORITY_REVOKED_AFTER_DISPATCH";
      } else {
        return { status: "FAIL", reason_code: "ILLEGAL_EFFECT_TRANSITION", effect_count: effectCount };
      }
    }
  }
  return { status, reason_code: reasonCode, effect_count: effectCount };
}

export function analyzeCapacity(profile, item) {
  if (item.unit !== profile.limits.capacity_unit) return { status: "FAIL", reason_code: "CAPACITY_UNIT_MISMATCH", max_backlog: null };
  const lanes = ["observation", "settlement", "recovery"];
  const lengths = lanes.flatMap((lane) => [item.arrivals?.[lane]?.length, item.service?.[lane]?.length]);
  if (lengths.some((length) => !Number.isInteger(length)) || new Set(lengths).size !== 1 || lengths[0] < 1) {
    return { status: "FAIL", reason_code: "CAPACITY_SHAPE_MISMATCH", max_backlog: null };
  }
  const numericValues = [
    ...lanes.map((lane) => item.initial[lane]),
    ...lanes.map((lane) => profile.limits.buffers[lane]),
    ...lanes.flatMap((lane) => item.arrivals[lane]),
    ...lanes.flatMap((lane) => item.service[lane])
  ];
  if (numericValues.some((value) => !Number.isFinite(value) || value < 0)) {
    return { status: "FAIL", reason_code: "CAPACITY_VALUE_INVALID", max_backlog: null };
  }
  const backlog = structuredClone(item.initial);
  const maxima = structuredClone(item.initial);
  let exceeded = false;
  for (let step = 0; step < lengths[0]; step += 1) {
    for (const lane of lanes) {
      backlog[lane] = Math.max(0, backlog[lane] + item.arrivals[lane][step] - item.service[lane][step]);
      maxima[lane] = Math.max(maxima[lane], backlog[lane]);
      if (backlog[lane] > profile.limits.buffers[lane]) exceeded = true;
    }
  }
  return { status: exceeded ? "FAIL" : "PASS", reason_code: exceeded ? "CAPACITY_EXCEEDED" : "CAPACITY_WITHIN_BOUNDS", max_backlog: maxima };
}

export function evaluateFixture(fixture) {
  return {
    batch_cases: fixture.batch_cases.map((item) => ({ case_id: item.case_id, result: analyzeBatch(fixture.profile, item) })),
    effect_cases: fixture.effect_cases.map((item) => ({ case_id: item.case_id, result: analyzeEffect(fixture.profile, item) })),
    capacity_cases: fixture.capacity_cases.map((item) => ({ case_id: item.case_id, result: analyzeCapacity(fixture.profile, item) }))
  };
}

export { canonical };

File: research/stp-v1.2/tools/concurrency.py
import itertools
import json
import pathlib
import sys


def canonical(value):
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"), sort_keys=True)


def same_set(left, right):
    return sorted(left) == sorted(right)


def validate_delta(profile, delta):
    if (delta["subject_id"] != profile["subject_id"] or delta["policy_id"] != profile["policy_id"] or
            delta["policy_generation"] != profile["policy_generation"]):
        return "POLICY_BINDING_MISMATCH"
    if not same_set(delta["causal_parents"], profile["causal_frontier"]):
        return "CAUSAL_FRONTIER_MISMATCH"
    return None


def predicate_holds(values, predicate):
    present = predicate["field"] in values
    actual = values.get(predicate["field"])
    operator, expected = predicate["operator"], predicate["value"]
    if operator == "EXISTS":
        return present == expected
    if not present:
        return False
    if operator == "EQ":
        return canonical(actual) == canonical(expected)
    if operator == "NE":
        return canonical(actual) != canonical(expected)
    if not isinstance(actual, (int, float)) or isinstance(actual, bool) or not isinstance(expected, (int, float)) or isinstance(expected, bool):
        return False
    return {
        "LT": actual < expected,
        "LTE": actual <= expected,
        "GT": actual > expected,
        "GTE": actual >= expected,
    }.get(operator, False)


def predicates_hold(values, delta):
    return all(predicate_holds(values, predicate) for predicate in delta.get("predicate_reads", []))


def run_schedule(profile, base_state, schedule):
    values = dict(base_state.get("values", {}))
    accepted, refused, effect_trace = [], [], []
    for delta in schedule:
        if not predicates_hold(values, delta):
            refused.append(delta["delta_id"])
            continue
        candidate = dict(values)
        candidate.update(delta.get("writes", {}))
        for key, increment in delta.get("increments", {}).items():
            candidate[key] = candidate.get(key, 0) + increment
        violated = any(isinstance(limit, (int, float)) and not isinstance(limit, bool) and
                       isinstance(candidate.get(key), (int, float)) and candidate[key] > limit
                       for key, limit in profile["limits"].items())
        if violated:
            refused.append(delta["delta_id"])
            continue
        values = candidate
        accepted.append(delta["delta_id"])
        effect_trace.extend(effect["effect_id"] for effect in delta.get("effects", []))
    return {
        "projection": {"values": values, "accepted": sorted(accepted), "refused": sorted(refused)},
        "effect_trace": effect_trace,
    }


def schedule_classes(profile, base_state, deltas):
    runs = [run_schedule(profile, base_state, schedule) for schedule in itertools.permutations(deltas)]
    return {
        "schedule_count": len(runs),
        "projection_count": len({canonical(run["projection"]) for run in runs}),
        "effect_count": len({canonical(run["effect_trace"]) for run in runs}),
    }


def analyze_batch(profile, item):
    if item["termination_evidence"] != "FINITE_ENUMERATION":
        return {"status": "UNKNOWN", "reason_code": "TERMINATION_NOT_PROVEN", "schedule_count": 0, "pairwise_all_commute": False}
    for delta in item["deltas"]:
        rejection = validate_delta(profile, delta)
        if rejection:
            return {"status": "FAIL", "reason_code": rejection, "schedule_count": 0, "pairwise_all_commute": False}
    if any(not predicates_hold(item["base_state"].get("values", {}), delta) for delta in item["deltas"]):
        return {"status": "FAIL", "reason_code": "PRECONDITION_FAILED", "schedule_count": 0, "pairwise_all_commute": False}
    full = schedule_classes(profile, item["base_state"], item["deltas"])
    pairs = list(itertools.combinations(item["deltas"], 2))
    pairwise_all = all((analysis := schedule_classes(profile, item["base_state"], list(pair)))["projection_count"] == 1 and
                       analysis["effect_count"] == 1 for pair in pairs)
    if full["projection_count"] > 1:
        return {
            "status": "FAIL",
            "reason_code": "PAIRWISE_INSUFFICIENT" if len(item["deltas"]) > 2 and pairwise_all else "ORDER_REQUIRED",
            "schedule_count": full["schedule_count"],
            "pairwise_all_commute": pairwise_all,
        }
    if full["effect_count"] > 1:
        return {"status": "FAIL", "reason_code": "EFFECT_ORDER_REQUIRED", "schedule_count": full["schedule_count"], "pairwise_all_commute": False}
    return {"status": "PASS", "reason_code": "COMMUTES", "schedule_count": full["schedule_count"], "pairwise_all_commute": pairwise_all}


def analyze_effect(profile, item):
    status, reason, effect_count = "UNKNOWN", "NO_EVENT", 0
    effects, keys = {}, {}
    for index, event in enumerate(item["events"]):
        expected_sequence = index + 1
        expected_previous = None if index == 0 else index
        if event["sequence"] != expected_sequence or event["previous_sequence"] != expected_previous:
            return {"status": "FAIL", "reason_code": "NONCONTIGUOUS_EFFECT_SEQUENCE", "effect_count": effect_count}
        if event["type"] != "OBSERVE" and event["policy_generation"] != profile["policy_generation"]:
            return {"status": "FAIL", "reason_code": "POLICY_BINDING_MISMATCH", "effect_count": effect_count}
        key_owner = keys.get(event["idempotency_key"])
        if key_owner is not None and key_owner != event["effect_id"]:
            return {"status": "FAIL", "reason_code": "EFFECT_ID_MISMATCH", "effect_count": effect_count}
        effect = effects.get(event["effect_id"])
        if effect is None and event["type"] != "PREPARE":
            return {"status": "FAIL", "reason_code": "EFFECT_ID_MISMATCH" if effects else "ILLEGAL_EFFECT_TRANSITION", "effect_count": effect_count}
        event_type = event["type"]
        if event_type == "PREPARE":
            if effect is not None:
                return {"status": "FAIL", "reason_code": "ILLEGAL_EFFECT_TRANSITION", "effect_count": effect_count}
            effect = {"state": "PREPARED", "idempotency_key": event["idempotency_key"], "dispatch_step": None,
                      "duplicate_pending": False, "outcomes": set()}
            effects[event["effect_id"]] = effect
            keys[event["idempotency_key"]] = event["effect_id"]
            status, reason = "UNKNOWN", "PREPARED_NOT_FINAL"
        elif effect["idempotency_key"] != event["idempotency_key"]:
            return {"status": "FAIL", "reason_code": "EFFECT_ID_MISMATCH", "effect_count": effect_count}
        elif event_type == "DISPATCH":
            if effect["state"] == "PREPARED":
                effect["state"], effect["dispatch_step"] = "DISPATCHED", event["effect_step"]
                effect_count += 1
                status, reason = "UNKNOWN", "DISPATCHED_NOT_FINAL"
            elif effect["state"] == "DISPATCHED":
                if event["effect_step"] - effect["dispatch_step"] <= profile["effect_policy"]["dedupe_retention_steps"]:
                    effect["duplicate_pending"] = True
                    status, reason = "UNKNOWN", "DUPLICATE_SUPPRESSION_UNVERIFIED"
                else:
                    effect_count += 1
                    effect["state"] = "DUPLICATE_RISK"
                    status, reason = "FAIL", "DUPLICATE_RISK"
            else:
                return {"status": "FAIL", "reason_code": "ILLEGAL_EFFECT_TRANSITION", "effect_count": effect_count}
        elif event_type == "DEDUPE_CONFIRM":
            if effect["state"] != "DISPATCHED" or not effect["duplicate_pending"]:
                return {"status": "FAIL", "reason_code": "ILLEGAL_EFFECT_TRANSITION", "effect_count": effect_count}
            effect["duplicate_pending"] = False
            status, reason = "PASS", "DUPLICATE_DEDUPED"
        elif event_type == "TIMEOUT":
            if effect["state"] != "DISPATCHED":
                return {"status": "FAIL", "reason_code": "ILLEGAL_EFFECT_TRANSITION", "effect_count": effect_count}
            effect["state"] = "EFFECT_UNKNOWN"
            status, reason = "UNKNOWN", "EFFECT_UNKNOWN"
        elif event_type == "OBSERVE":
            if effect["state"] not in {"DISPATCHED", "EFFECT_UNKNOWN", "OBSERVED"}:
                return {"status": "FAIL", "reason_code": "ILLEGAL_EFFECT_TRANSITION", "effect_count": effect_count}
            if event["policy_generation"] != profile["policy_generation"]:
                effect["state"] = "OBSERVATION_STALE"
                status, reason = "UNKNOWN", "OBSERVATION_STALE"
            else:
                effect["outcomes"].add(event["outcome"])
                if len(effect["outcomes"]) > 1:
                    effect["state"] = "OBSERVATION_CONFLICT"
                    status, reason = "UNKNOWN", "OBSERVATION_CONFLICT"
                else:
                    effect["state"] = "OBSERVED"
                    status, reason = "UNKNOWN", "OBSERVED_NOT_SETTLED"
        elif event_type == "SETTLE":
            if effect["state"] != "OBSERVED":
                return {"status": "FAIL", "reason_code": "ILLEGAL_EFFECT_TRANSITION", "effect_count": effect_count}
            effect["state"] = "SETTLED"
            status, reason = "PASS", "SETTLED"
        elif event_type == "REVOKE":
            if effect["state"] == "PREPARED":
                effect["state"] = "REFUSED"
                status, reason = "FAIL", "AUTHORITY_REVOKED_BEFORE_DISPATCH"
            elif effect["state"] in {"DISPATCHED", "EFFECT_UNKNOWN", "OBSERVED"}:
                effect["state"] = "EFFECT_UNKNOWN"
                status, reason = "UNKNOWN", "AUTHORITY_REVOKED_AFTER_DISPATCH"
            else:
                return {"status": "FAIL", "reason_code": "ILLEGAL_EFFECT_TRANSITION", "effect_count": effect_count}
    return {"status": status, "reason_code": reason, "effect_count": effect_count}


def analyze_capacity(profile, item):
    if item["unit"] != profile["limits"]["capacity_unit"]:
        return {"status": "FAIL", "reason_code": "CAPACITY_UNIT_MISMATCH", "max_backlog": None}
    lanes = ["observation", "settlement", "recovery"]
    lengths = [len(item.get("arrivals", {}).get(lane, [])) for lane in lanes] + [len(item.get("service", {}).get(lane, [])) for lane in lanes]
    if not lengths or len(set(lengths)) != 1 or lengths[0] < 1:
        return {"status": "FAIL", "reason_code": "CAPACITY_SHAPE_MISMATCH", "max_backlog": None}
    numeric_values = ([item["initial"][lane] for lane in lanes] + [profile["limits"]["buffers"][lane] for lane in lanes] +
                      [value for lane in lanes for value in item["arrivals"][lane]] +
                      [value for lane in lanes for value in item["service"][lane]])
    if any(not isinstance(value, (int, float)) or isinstance(value, bool) or value < 0 for value in numeric_values):
        return {"status": "FAIL", "reason_code": "CAPACITY_VALUE_INVALID", "max_backlog": None}
    backlog, maxima = dict(item["initial"]), dict(item["initial"])
    exceeded = False
    for step in range(lengths[0]):
        for lane in lanes:
            backlog[lane] = max(0, backlog[lane] + item["arrivals"][lane][step] - item["service"][lane][step])
            maxima[lane] = max(maxima[lane], backlog[lane])
            exceeded = exceeded or backlog[lane] > profile["limits"]["buffers"][lane]
    return {"status": "FAIL" if exceeded else "PASS", "reason_code": "CAPACITY_EXCEEDED" if exceeded else "CAPACITY_WITHIN_BOUNDS", "max_backlog": maxima}


def evaluate_fixture(fixture):
    return {
        "batch_cases": [{"case_id": item["case_id"], "result": analyze_batch(fixture["profile"], item)} for item in fixture["batch_cases"]],
        "effect_cases": [{"case_id": item["case_id"], "result": analyze_effect(fixture["profile"], item)} for item in fixture["effect_cases"]],
        "capacity_cases": [{"case_id": item["case_id"], "result": analyze_capacity(fixture["profile"], item)} for item in fixture["capacity_cases"]],
    }


if __name__ == "__main__":
    fixture = json.loads(pathlib.Path(sys.argv[1]).read_text(encoding="utf-8"))
    print(canonical(evaluate_fixture(fixture)))
